src/control/client/config.go

```diff
@@ -0,0 +1,195 @@
+	Path          string
+	ext           External
+}
+
+// newDefaultConfiguration creates a new instance of configuration struct
+// populated with defaults.
+func newDefaultConfiguration(ext External) Configuration {
+	return Configuration{
+		SystemName:    defaultSystemName,
+		AccessPoints:  []string{"localhost"},
+		Port:          defaultPort,
+		HostList:      []string{"localhost:10001"},
+		RuntimeDir:    defaultRuntimeDir,
+		HostFile:      "",
+		Cert:          "",
+		Key:           "",
+		LogFile:       defaultLogFile,
+		LogFileFormat: "",
+		Path:          defaultConfigPath,
+		ext:           ext,
+	}
+}
+
+// ProcessConfigFile loads a configuration file from the path given,
+// or from the default location if none is provided.  It returns a populated
+// Configuration struct based upon the default values and any config file overrides.
+func ProcessConfigFile(ConfigPath string) (Configuration, error) {
+	config := NewConfiguration()
+	if ConfigPath != "" {
+		log.Debugf("Overriding default config path with: %s", ConfigPath)
+		config.Path = ConfigPath
+	}
+
+	if !filepath.IsAbs(config.Path) {
+		newPath, err := common.GetAbsInstallPath(config.Path)
+		if err != nil {
+			return config, err
+		}
+		config.Path = newPath
+	}
+
+	err := config.LoadConfig()
+	if err != nil {
+		log.Debugf("Unable to read the configuration file from: '%s'  Using default configuration.", config.Path)
+		return config, nil
+	}
+	log.Debugf("DAOS Client config read from %s", config.Path)
+
+	return config, nil
+}
+
+// LoadConfig reads the configuration file specified by Configuration.Path
+// and parses it.  Parsed values override any default values.
+func (c *Configuration) LoadConfig() error {
+	bytes, err := ioutil.ReadFile(c.Path)
+	if err != nil {
+		return err
+	}
+
+	if err = c.parse(bytes); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ApplyAgentCmdLineOverrides will overwrite Configuration values with any non empty
+// data provided, usually from the commandline.
+func (c *Configuration) ApplyAgentCmdLineOverrides(RuntimeDir string, LogFile string) error {
+
+	if RuntimeDir != "" {
+		log.Debugf("Overriding socket path from config file with %s", RuntimeDir)
+		c.RuntimeDir = RuntimeDir
+	}
+
+	if LogFile != "" {
+		log.Debugf("Overriding LogFile path from config file with %s", LogFile)
+		c.LogFile = LogFile
+	}
+
+	return nil
+}
+
+// ApplyDMGCmdLineOverrides will overwrite Configuration values with any non empty
+// data provided, usually from the commandline.
+func (c *Configuration) ApplyDMGCmdLineOverrides(Hostlist string) error {
+
+	if (len(Hostlist) > 0) {
+		hosts := strings.Split(Hostlist, ",")
+		log.Debugf("Overriding hostlist from config file with %s", hosts)
+		c.HostList = hosts
+	}
+
+	return nil
+}
+
+// ProcessEnvOverrides examines environment variables and applies the environment
+// variable value over an existing value if defined.
+func (c *Configuration) ProcessEnvOverrides() int {
```

Comment:
IMO this gets a little confusing with the value also defined in the config file. For any non-default value, the env variable will have to be set to run client apps. If someone sets up a service to start the agent automatically, it seems like they'd want to set up the env variable to be exported to all users globally, too. As I look at it, I guess I'm not sure of the use case where someone wants to use the config file value and has not set the env variable.

Reply:
As I understood it, the environment variable configuration was a legacy mode that we are moving away from.  But, we wanted to be able to process these environment variable values if the system was still providing them (for now).  My expectation is that we will be removing the environment variable overrides in the near future, similar to what Johann talked about today when he mentioned deprecating those variables on the server.  I think we would be in the clear if we said that we're not going to support environment variables at all.  But, Tom did ask for this feature originally, so let's see what he thinks about it as of now.

Reply:
This is kind of a weird case - the reason for having the env variable at all is because the client library (libdaos) requires it. It doesn't have any other way to figure out a non-default socket path, and needs the path to communicate with the agent. Unless we're having libdaos process the config file, too, we'll need that variable set if the socket path is anything but the default.

I guess the question is whether we want to use the env variable in the agent at all. Is it currently supported in master? If not maybe it's better not to start.

Reply:
I understand what you mean.  If libdaos requires the environment variable, then it doesn't make sense to make this configurable via config file or commandline override either.  Maybe Tom will have some fresh eyes and has a good perspective on what he wants to happen here.    At face value, it makes sense to remove the value from the config file and the command line, and actually generate an error if the environment variable is left undefined.

Reply:
I think I get where Kris is coming from; libdaos and agent apps should use same default sock location, if non-default location is required, both apps should read from same env var. Therefore no sock location should be in client config and agent should check env var before setting default sock location (.or do we rely on it being passed to agent in cli opts Rather than reading from env?) Does that align with your thoughts @kjacque ?

Reply:
I'm okay with either approach (config file or env var), it's just the combination that makes my head hurt a little. If we do want to use the config file for this, I think we should ignore the env variable completely in the agent. Letting the environment override values canonized in a config file feels weird to me.

Reply:
Because libdaos uses the environment variable today, and this still seems reasonable, my suggestion is that we do not specify the socket path through any other means.  That would mean, remove it from the config file and remove it from the commandline options.  What do you say?

Reply:
Using the environment variable as the source for the path is fine with me. I'd lean toward leaving the command line override in place - not sure if it is currently used in CI.

Reply:
If we use the commandline override, how will the libdaos stay in sync if it only uses the environment variable?  Did you intend for the client to then export the value to the environment so they stay in sync (assuming libdaos loads after)?  Or, do you intend for these to be able to diverge?  Either is OK with me .. I'm just not clear on the use model any longer for allowing an override in this particular case.

Reply:
My thought was if someone starts the agent before exporting the env variable. Not sure if there is another use case. Yes, they would need to set the env variable anyway to use libdaos, so maybe it's moot. But at the very least we need to verify CI tests are updated accordingly if the command line option is removed from the agent.

I'd suggest running this idea by @johannlombardi to ensure it makes sense to him, too.

Reply:
We had a discussion with the team @liw @dpquigl @kjacque @tanabarr @amjustin13 @johannlombardi  and Johann decided:  

1.  the socket path shall remain a variable in the daos.yml config file even though it is only used by the agent
2.  the agent may continue to specify a command line socket path value that overrides the config file
3.  the config.go shall not use the environment variable DAOS_AGENT_DRPC_DIR to **override** the socket path.  
4.  config.go may read DAOS_AGENT_DRPC_DIR to examine its value and print a warning if it conflicts.  This conflict will not be considered a fatal error.

Johann left the meeting prior to hearing feedback from @liw that the socket path should not be in the config file because it is agent specific, and believes that the config.go should use DAOS_AGENT_DRPC_DIR env var as an override of the config file and command line.  

The implication of Johann's feedback is: to successfully override this value, the user must modify the DAOS_AGENT_DRPC_DIR env for libdaos and **separately** configure the config file or command line.    The rationale is that libdaos will not be asked to parse any config file.  

If there are any inaccuracies in the feedback captured, please respond here.  In the interest of wrapping this up, if you disagree with Johann's feedback (1-4 above), please respond here by your end of day 5/10 so we can make progress resolving this.

Reply:
I'm fine with making progress based on 1-4 above.

Reply:
Going with 1-4 as above.